### PR TITLE
HHH-14416 test cases for re-saving a deleted entity

### PR DESCRIPTION
**Pull request by karge-itestra**
https://hibernate.atlassian.net/browse/HHH-14416

### CODE REVIEW

**Comment by karge-itestra** [comment]
I have added an additional test case showing that an entity entry that is created by `em.persist()` is created with `isExistsInDatabase() == false` instead of evaluating whether the entity is already in the database. This happens although the entity manager determines correctly that there was a delete statement that has not been flushed yet (so the entity must exist in the database)

File: hibernate-core/src/test/java/org/hibernate/event/ReSaveReferencedDeletedEntity.java [hunk 1]
```diff
@@ -0,0 +1,95 @@
+package org.hibernate.event;
+
+import org.hibernate.testing.TestForIssue;
+import org.hibernate.testing.junit4.BaseCoreFunctionalTestCase;
+import org.junit.Test;
+
+import javax.persistence.*;
```
Comment (gavinking) [inline hibernate-core/src/test/java/org/hibernate/event/ReSaveReferencedDeletedEntity.java:7]:
```suggestion
import jakarta.persistence.*;
```

File: hibernate-core/src/test/java/org/hibernate/event/ReSaveReferencedDeletedEntityJPA.java [hunk 2]
```diff
@@ -0,0 +1,100 @@
+package org.hibernate.event;
+
+import org.hibernate.Session;
+import org.hibernate.jpa.test.BaseEntityManagerFunctionalTestCase;
+import org.hibernate.testing.TestForIssue;
+import org.junit.Test;
+
+import javax.persistence.*;
```
Comment (gavinking) [inline hibernate-core/src/test/java/org/hibernate/event/ReSaveReferencedDeletedEntityJPA.java:8]:
```suggestion
import jakarta.persistence.*;
```

File: hibernate-core/src/test/java/org/hibernate/event/ReSaveReferencedDeletedEntityJPA.java [hunk 3]
```diff
@@ -0,0 +1,100 @@
+package org.hibernate.event;
+
+import org.hibernate.Session;
+import org.hibernate.jpa.test.BaseEntityManagerFunctionalTestCase;
+import org.hibernate.testing.TestForIssue;
```
Comment (gavinking) [inline hibernate-core/src/test/java/org/hibernate/event/ReSaveReferencedDeletedEntityJPA.java:5]:
```suggestion
import org.hibernate.testing.orm.junit.JiraKey;
```

File: hibernate-core/src/test/java/org/hibernate/event/ReSaveReferencedDeletedEntityJPA.java [hunk 4]
```diff
@@ -0,0 +1,100 @@
...
+
+        em.detach( parent );
+
+        em.persist( parent );
+
+        em.refresh( child );
+    }
+
+    @Test
+    @TestForIssue(jiraKey = "HHH-14416")
```
Comment (gavinking) [inline hibernate-core/src/test/java/org/hibernate/event/ReSaveReferencedDeletedEntityJPA.java:43]:
```suggestion
    @JiraKey("HHH-14416")
```

File: hibernate-core/src/test/java/org/hibernate/event/ReSaveReferencedDeletedEntityJPA.java [hunk 5]
```diff
@@ -0,0 +1,100 @@
...
+import jakarta.persistence.*;
+
+public class ReSaveReferencedDeletedEntityJPA extends BaseEntityManagerFunctionalTestCase {
+    @Override
+    protected Class<?>[] getAnnotatedClasses() {
+        return new Class<?>[] { Child.class, Parent.class };
+    }
+
+    @Test
+    @TestForIssue(jiraKey = "HHH-14416")
```
Comment (gavinking) [inline hibernate-core/src/test/java/org/hibernate/event/ReSaveReferencedDeletedEntityJPA.java:17]:
```suggestion
    @ JiraKey("HHH-14416")
```

File: hibernate-core/src/test/java/org/hibernate/event/ReSaveReferencedDeletedEntity.java [hunk 6]
```diff
@@ -0,0 +1,95 @@
+package org.hibernate.event;
+
+import org.hibernate.testing.TestForIssue;
```
Comment (gavinking) [inline hibernate-core/src/test/java/org/hibernate/event/ReSaveReferencedDeletedEntity.java:3]:
```suggestion
import org.hibernate.testing.orm.junit.JiraKey;
```

File: hibernate-core/src/test/java/org/hibernate/event/ReSaveReferencedDeletedEntity.java [hunk 7]
```diff
@@ -0,0 +1,95 @@
...
+import static org.hibernate.testing.transaction.TransactionUtil.doInHibernate;
+
+public class ReSaveReferencedDeletedEntity extends BaseCoreFunctionalTestCase {
+    @Override
+    protected Class<?>[] getAnnotatedClasses() {
+        return new Class<?>[] { Child.class, Parent.class };
+    }
+
+    @Test
+    @TestForIssue(jiraKey = "HHH-14416")
```
Comment (gavinking) [inline hibernate-core/src/test/java/org/hibernate/event/ReSaveReferencedDeletedEntity.java:18]:
```suggestion
    @JiraKey("HHH-14416")
```

File: hibernate-core/src/test/java/org/hibernate/event/ReSaveReferencedDeletedEntity.java [hunk 8]
```diff
@@ -0,0 +1,95 @@
...
+            session.saveOrUpdate( child );
+
+            session.remove(child);
+
+            session.save(child);
+        } );
+    }
+
+    @Test
+    @TestForIssue(jiraKey = "HHH-14416")
```
Comment (gavinking) [inline hibernate-core/src/test/java/org/hibernate/event/ReSaveReferencedDeletedEntity.java:36]:
```suggestion
    @JiraKey("HHH-14416")
```

**Comment by gavinking** [comment]
The tests proposed here contained several user misunderstandings, but I did find something interesting, so moving to #9316.

**Pull request closed without merging**